### ISSUE TITLE
Parallelizing via ncores in snp_asGeneticPos()

### ISSUE THREAD

**Issue by char4816** [privefl/bigsnpr#117]
After using ldpred1 and SCT, I'm really excited about the improvements you have made in ldpred2. Thank you for your hard work!  As I am working through the vignette I encountered an issue at the point of:

`POS2 <- snp_asGeneticPos(CHR, POS, dir = "tmp-data", ncores = 3)`

I am operating on an HPC with slurm workload manager.  If I submitted my R script via sbatch, the script hangs at this line (forever).  I eventually was able to solve the issue by just removing the ncores from the function call:

`POS2 <- snp_asGeneticPos(CHR, POS, dir = "tmp-data")`

I played with inputting different integers for the ncores parameter, but it seems there might be a bug in the parallelization within this function? All of the other functions that parallelize across cores have been working for me so far.

Not sure if this is a problem could be specific to my computing resources, but just wanted to throw it out there in case I did in fact come across a bug.

Thanks again!
Chris

**Comment by privefl** [comment]
Thanks for reporting, but I'm quite surprised.
The parallelization used is very standard, and I never had any problem with this (I'm also using some HPC cluster with slurm).

How long is "forever"? How many cores have you tried with how many you asked for?

**Comment by char4816** [comment]
I had tried many different computing setups and even increased up to 16 cores with 500GB RAM.
What was really throwing me for a loop the past few days is that if I ran  the script line by line with 
`srun -p bigmem -n 16 -t 0-8:00 --mem=500GB --pty R`
then it would work just fine.

However, if I submitted the script as a sbatch using
`sbatch -p bigmem -n 16 -t 0-8:00 --mem=500GB R CMD BATCH c_ldpred2_adrn_trial.R`
then it would get hung at the snp_asGeneticPos function.

But I just had an idea and I think I got to the bottom of it.  

On our HPC when we do R parallelization with parallel::makeCluster, we've found that the "socket" method for makeCluster is unsafe and can crash nodes, but the "fork" method seems to be well behaved.  So I tend to try and use "fork."

To force "fork" to be the default instead of "socket," I would just hijack the default of the parallel::makeCluster function at the top of my script with something like

```
hijack <- function (FUN, ...) {
    .FUN <- FUN
    args <- list(...)
    invisible(lapply(seq_along(args), function(i) {
        formals(.FUN)[[names(args)[i]]] <<- args[[i]]
    }))
    .FUN
}

tmp <- hijack(parallel::makeCluster, type = "FORK")
assignInNamespace("makeCluster",tmp,ns="parallel")
```

I just figured out that snp_asGeneticPos runs great if I **don't** hijack makeCluster (so type="PSOCK"), but gets stuck if I **do** hijack makeCluster (so type="FORK")

So it seems that for some reason the snp_asGeneticPos function becomes hung up when submitted via sbatch and type="FORK", but all of the other functions in your package are okay when run via sbatch _or_ srun with type="FORK".  

Long story short, maybe adding a parallelization parameter for the method used in makeCluster could be a future improvement to ldpred2?

**Comment by privefl** [comment]
I thought sockets were always safer than fork.

I guess I could use the same option as in {bigstatsr} (cf. e.g. https://github.com/privefl/bigstatsr/blob/master/R/apply-parallelize.R#L54).

**Comment by char4816** [comment]
That would be great if you find the time to make that an option. I'll close this issue now.

Thank you for your awesome work, quick reply, and support,
Chris
